x-pack/solutions/observability/plugins/streams/server/lib/streams/client.ts

```diff
@@ -739,38 +791,124 @@ export class StreamsClient {
     });
   }
 
-  async getAncestors(name: string): Promise<WiredStreamDefinition[]> {
-    const ancestorIds = getAncestors(name);
+  /**
+   * Returns the unwired root stream for a wired stream.
+   *
+   * For a stream wired to logs, this returns undefined.
+   * For a stream wired to an unwired classic stream, this returns the classic stream.
+   *
+   * This is done by fetching all classic stream definitions that start with the first dot part of the dataset
+   * and looking for the longest match (in terms of numbers of dots in the dataset).
+   */
+  async getUnwiredRootForStream(
+    definition: WiredStreamDefinition
+  ): Promise<StreamDefinition | undefined> {
+    let classicRoot: StreamDefinition | undefined;
+    let classicRootSegmentLength = 0;
+    const streamName = parseStreamName(definition.name);
+    if (streamName.type === 'dsns') {
+      const streams = await this.getStreamDefinitions({
+        query: {
+          bool: {
+            filter: [
+              {
+                prefix: {
+                  name: `${streamName.datastreamType}-${streamName.datasetSegments[0]}`,
+                },
+              },
+            ],
+          },
+        },
+      });
+
+      streams
+        .filter((candidateStream) => !isWiredStream(candidateStream))
+        .forEach((candidateStream) => {
+          const candidateStreamName = parseStreamName(candidateStream.name);
+          // The longest unwired stream with a matching prefix is the classic root
```

Comment:
What happens if you have two classic streams like:
```
logs-mydata
logs-mydata.subset
```

And then fork both of them to some wired children? is that allowed? 
I _think_ this code will try to look for `logs-mydata` as the prefix in both cases and then for both cases return the longer stream name?

Reply:
it will return the longest that matches. So you should be able to have classic streams `logs-mydata-default` and `logs-mydata.subset-default`, then fork both of them into `logs-mydata.myfork-default` and `logs-mydata.subset.myfork-default` respectively.

I didn't test what happens when you try to fork `logs-mydata.subset-default` from `logs-mydata-default` even though both are classic streams already, that should be fixed as part of https://github.com/elastic/streams-program/issues/26

Reply:
However, it's a good point, I will add a test for this case

Reply:
> I didn't test what happens when you try to fork logs-mydata.subset-default from logs-mydata-default even though both are classic streams already

You get this legitimate error `Cannot change unwired stream to wired stream`